File: iosApp/iosApp/Views/ToolbarIcons.swift
/// Custom, hand-drawn toolbar glyphs shared across the termtastic clients.
///
/// Both shapes here are drawn from scratch with SwiftUI `Path`s using the same
/// 24-unit coordinate geometry as their Android (`LayoutGridIcon` / `PlusIcon`
/// in `ToolbarIcons.kt`) and web (`ICON_LAYOUT` / `ICON_NEW_TAB` inline SVGs in
/// darkness-toolkit) counterparts, so every platform shows an identical mark.
/// They replace the SF Symbols `rectangle.3.group` and `plus` in the Sessions
/// toolbar (`TreeView`).

import SwiftUI

/// The "layout presets" shape: one large pane on the left with two smaller
/// panes stacked on the right.
/// Inspired by the SF Symbol `rectangle.3.group`. Stroked (not filled) so it
/// reads as an outline glyph matching the other toolbar symbols.
private struct LayoutGridShape: Shape {
    /// Builds three rounded-rectangle panes scaled into `rect` from the shared
    /// 24-unit design space.
    ///
    /// - Parameter rect: the bounds the glyph is laid out in.
    /// - Returns: a `Path` containing the three pane outlines.
    func path(in rect: CGRect) -> Path {
        let u = min(rect.width, rect.height) / 24
        func pane(_ x: CGFloat, _ y: CGFloat, _ w: CGFloat, _ h: CGFloat) -> Path {
            Path(
                roundedRect: CGRect(
                    x: rect.minX + x * u,
                    y: rect.minY + y * u,
                    width: w * u,
                    height: h * u
                ),
                cornerRadius: 1.2 * u
            )
        }
        var path = Path()
        path.addPath(pane(4, 5, 8.5, 14))
        path.addPath(pane(14.5, 5, 5.5, 6))
        path.addPath(pane(14.5, 13, 5.5, 6))
        return path
    }
}

/// Toolbar-ready "layout" glyph. Takes its colour from the surrounding
/// `.foregroundStyle`, exactly like an SF Symbol would.
struct LayoutGridIcon: View {
    var body: some View {
        GeometryReader { geo in
            let u = min(geo.size.width, geo.size.height) / 24
            LayoutGridShape()
                .stroke(style: StrokeStyle(lineWidth: 1.5 * u, lineJoin: .round))
        }
        .frame(width: 22, height: 22)
    }
}

/// A clean "+" shape with rounded line caps. Inspired by the SF Symbol `plus`,
/// drawn from scratch so it matches the Android/web add buttons exactly.
private struct PlusShape: Shape {
    /// Builds the two crossing strokes scaled into `rect` from the shared
    /// 24-unit design space.
    ///
    /// - Parameter rect: the bounds the glyph is laid out in.
    /// - Returns: a `Path` containing the vertical and horizontal strokes.
    func path(in rect: CGRect) -> Path {
        let u = min(rect.width, rect.height) / 24
        var path = Path()
        path.move(to: CGPoint(x: rect.minX + 12 * u, y: rect.minY + 5 * u))
        path.addLine(to: CGPoint(x: rect.minX + 12 * u, y: rect.minY + 19 * u))
        path.move(to: CGPoint(x: rect.minX + 5 * u, y: rect.minY + 12 * u))
        path.addLine(to: CGPoint(x: rect.minX + 19 * u, y: rect.minY + 12 * u))
        return path
    }
}

/// Toolbar-ready "add" glyph. Takes its colour from the surrounding
/// `.foregroundStyle`, exactly like an SF Symbol would.
struct PlusIcon: View {
    var body: some View {
        GeometryReader { geo in
            let u = min(geo.size.width, geo.size.height) / 24
            PlusShape()
                .stroke(style: StrokeStyle(lineWidth: 2 * u, lineCap: .round))
        }
        .frame(width: 22, height: 22)
    }
}

File: androidApp/src/main/kotlin/se/soderbjorn/termtastic/android/ui/ToolbarIcons.kt
/*
 * Custom, hand-drawn toolbar glyphs shared across the termtastic clients.
 *
 * Both icons in this file are drawn from scratch with a [Canvas] using the
 * same 24-unit coordinate geometry as their iOS (`LayoutGridIcon` /
 * `PlusIcon` in `ToolbarIcons.swift`) and web (`ICON_LAYOUT` / `ICON_NEW_TAB`
 * inline SVGs in darkness-toolkit) counterparts, so every platform shows an
 * identical mark. They replace the previous Material icons (`Icons.Outlined.
 * Dashboard` and `Icons.Filled.Add`) in the Sessions toolbar.
 */
package se.soderbjorn.termtastic.android.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.size
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/**
 * The "layout presets" glyph: one large pane on the left with two smaller
 * panes stacked on the right.
 *
 * Drawn from scratch but inspired by the iOS SF Symbol `rectangle.3.group`.
 * Called by `TreeScreen`'s toolbar in overview mode where it opens the layout
 * preset sheet; replaces the former `Icons.Outlined.Dashboard`.
 *
 * @param contentDescription accessibility label for the glyph (e.g. "Layout").
 * @param tint stroke colour for the three pane outlines.
 * @param modifier optional layout modifier; defaults to a 24.dp box matching
 *   the standard Material icon size.
 * @see PlusIcon
 */
@Composable
fun LayoutGridIcon(
    contentDescription: String,
    tint: Color,
    modifier: Modifier = Modifier,
) {
    Canvas(
        modifier
            .size(24.dp)
            .semantics {
                this.contentDescription = contentDescription
                role = Role.Image
            },
    ) {
        val u = size.minDimension / 24f
        val stroke = Stroke(width = 1.5f * u)
        val radius = CornerRadius(1.2f * u, 1.2f * u)
        fun pane(x: Float, y: Float, w: Float, h: Float) =
            drawRoundRect(
                color = tint,
                topLeft = Offset(x * u, y * u),
                size = Size(w * u, h * u),
                cornerRadius = radius,
                style = stroke,
            )
        pane(4f, 5f, 8.5f, 14f)
        pane(14.5f, 5f, 5.5f, 6f)
        pane(14.5f, 13f, 5.5f, 6f)
    }
}

/**
 * A clean "+" glyph with rounded line caps.
 *
 * Drawn from scratch but inspired by the iOS SF Symbol `plus`. Called by
 * `TreeScreen`'s toolbar "Add" menu button; replaces the former
 * `Icons.Filled.Add` so the add affordance matches across platforms.
 *
 * @param contentDescription accessibility label for the glyph (e.g. "Add").
 * @param tint stroke colour for the two crossing strokes.
 * @param modifier optional layout modifier; defaults to a 24.dp box matching
 *   the standard Material icon size.
 * @see LayoutGridIcon
 */
@Composable
fun PlusIcon(
    contentDescription: String,
    tint: Color,
    modifier: Modifier = Modifier,
) {
    Canvas(
        modifier
            .size(24.dp)
            .semantics {
                this.contentDescription = contentDescription
                role = Role.Image
            },
    ) {
        val u = size.minDimension / 24f
        val stroke = Stroke(width = 2f * u, cap = StrokeCap.Round)
        drawLine(tint, Offset(12f * u, 5f * u), Offset(12f * u, 19f * u), strokeWidth = stroke.width, cap = stroke.cap)
        drawLine(tint, Offset(5f * u, 12f * u), Offset(19f * u, 12f * u), strokeWidth = stroke.width, cap = stroke.cap)
    }
}
